Initial support for EFR32MG24

fkokosinski opened the pull request:
This PR introduces initial support for the BRD2601B (Mighty Gecko Radio Board) board. It consists of:
* EFR32MG24 SoC support
* BRD2601B board support
* GPIO support
* UART support

msierszulski committed:
soc: silabs_exx32: Add support for SiLabs efr32mg24 SoC

This commit adds support for Silicon Labs EFR32MG24 SoC.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

msierszulski committed:
boards: Add support for SiLabs efr32xg24_dk2601b board

This commit adds support for Silicon Labs efr32xg24_dk2601b board.

Signed-off-by: Mateusz Sierszulski <msierszulski@antmicro.com>

lpawelcz committed:
drivers: gpio: gecko: add support for efr32xg24

Add defined() check for EFR32MG24

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

galak commented on boards/arm/efr32xg24_dk2601b/efr32xg24_dk2601b.yaml:
```diff
@@ -0,0 +1,17 @@
+toolchain:
+  - zephyr
+  - gnuarmemb
+  - xtools
```
drop xtools, its deprecated

fkokosinski replied:
Done

galak commented on soc/arm/silabs_exx32/Kconfig:
```diff
@@ -140,6 +140,11 @@ config CRYPTO_ACC_GECKO_TRNG
 	  Enable Entropy driver based on the CRYPTO_ACC module for Silicon Labs
 	  Gecko chips.
 
+config SOC_GECKO_SEMAILBOX
```
what is using this? If nothing remove for now.

fkokosinski replied:
Good catch. This symbol isn't used right now. I've removed it for now, and will reintroduce it in another PR.

galak reviewed (dismissed):
in general LGTM.  A few minor questions and cleanups.

galak reviewed (commented):
What uses sl_device_init_hfxo_config.h?

fkokosinski committed:
tests: mark testcases with `pm` where `CONFIG_PM=y` is forced

This commit marks testcases that require working Power Managament with
the appropriate `pm` tag to allow proper testcase filtering in the board
YAML file.

Signed-off-by: Filip Kokosinski <fkokosinski@antmicro.com>

fkokosinski commented:
> What uses sl_device_init_hfxo_config.h?

Hi @galak. This file is used by the HAL here: https://github.com/zephyrproject-rtos/hal_silabs/blob/master/gecko/CMakeLists.txt#L36-L41. Similar setup is used by the `efr32bg_sltb010a` board: https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arm/efr32bg_sltb010a/sl_device_init_hfxo_config.h.

fkokosinski commented:
Hi @galak, please let me know if there is anything left for me to fix here before we can proceed with this PR.

lpawelcz committed:
drivers: uart: gecko: add support for efr32xg24

Add missing case when there is only one USART

Signed-off-by: Pawel Czarnecki <pczarnecki@antmicro.com>

fkokosinski commented on boards/arm/efr32xg24_dk2601b/board.c:
```diff
@@ -0,0 +1,71 @@
+#include "em_cmu.h"
+#endif
+
+LOG_MODULE_REGISTER(efr32xg24_dk2601b, LOG_LEVEL_DBG);
```
Changed to `CONFIG_BOARD_EFR32MG24_LOG_LEVEL`.

fkokosinski commented:
@tgorochowik thanks for the review, I've applied the changes you suggested. PTAL at this PR.

fkokosinski commented:
Hi @dcpleung, @mnkp. Could you check if the UART/GPIO changes (respectively) are fine with you? Thanks!

dcpleung reviewed (dismissed):
UART LGTM

fkokosinski commented:
Hi @tgorochowik, @dcpleung.

I had to force push to update the sha pointer to the `hal_silabs` repository after https://github.com/zephyrproject-rtos/hal_silabs/pull/22 got merged, dismissing your ACKs in the process. PTAL at this PR. Thanks!

mrfuchs commented on soc/arm/silabs_exx32/efr32mg24/soc.h:
```diff
@@ -0,0 +1,30 @@
+
+/**
+ * @file
+ * @brief Board configuration macros for the EFR32MG21 SoC
```
Shouldn't this read EFR32MG24?

fkokosinski replied:
Done

mrfuchs commented on soc/arm/silabs_exx32/efr32mg24/soc_pinmap.h:
```diff
@@ -0,0 +1,22 @@
+ */
+
+/** @file
+ * @brief Silabs EFR32MG21 MCU pin definitions.
```
Shouldn't this read EFR32MG24?

fkokosinski replied:
Done

mrfuchs commented on boards/arm/efr32xg24_dk2601b/doc/index.rst:
```diff
@@ -0,0 +1,154 @@
+Serial Port
+===========
+
+The EFR32MG24 SoC has two USARTs.
```
According to the EFR32xG24 Wireless SoC Reference Manual it has one USART and two EUSARTs.

mrfuchs reviewed (changes requested):
Some minor documentation nits, otherwise LGTM.

Was able to create a BRD4187C board based on this PR.

ihidchaos commented:
Hi, I would like to add PWM-led light support for brd2601b based on this PR. 
I made the following changes and verified them with the example samples/drivers/led_pwm. 
However, after brushing the firmware into the device, the led lights did not flicker on the light change. 
How do I change the code to make it work？

```patch
diff --git a/boards/arm/efr32xg24_dk2601b/efr32xg24_dk2601b.dts b/boards/arm/efr32xg24_dk2601b/efr32xg24_dk2601b.dts
index e7ff6d3555..cc26a4a385 100644
--- a/boards/arm/efr32xg24_dk2601b/efr32xg24_dk2601b.dts
+++ b/boards/arm/efr32xg24_dk2601b/efr32xg24_dk2601b.dts
@@ -7,6 +7,7 @@
 /dts-v1/;
 #include <silabs/efr32mg24b310f1536im48.dtsi>
 #include <silabs/efr32mg24-pinctrl.dtsi>
+#include <zephyr/dt-bindings/pwm/pwm.h>
 
 / {
 	model = "Silicon Labs BRD2601B (Mighty Gecko Radio Board)";
@@ -26,6 +27,7 @@
 		led2 = &blue_led;
 		sw0 = &button0;
 		sw1 = &button1;
+		pwm-led0 = &pwm_led0;
 	};
 
 	leds {
@@ -41,6 +43,15 @@
 		};
 	};
 
+	pwmleds {
+		compatible = "pwm-leds";
+		status = "okay";
+		pwm_led0: pwm_led0 {
+			label = "pwm_led0";
+			pwms = <&pwm0 0 PWM_MSEC(20) PWM_POLARITY_NORMAL>;
+		};
+	};
+
 	buttons {
 		compatible = "gpio-keys";
 		button0: button_0 {
@@ -69,6 +80,16 @@
 	status = "okay";
 };
 
+&timer0 {
+	status = "okay";
+
+	pwm0: pwm {
+		status = "okay";
+		pin-location = <GECKO_LOCATION(0) GECKO_PORT_B GECKO_PIN(0)>;
+		prescaler = <1024>;
+	};
+};
+
 &gpio {
 	status = "okay";
 };
diff --git a/drivers/pwm/pwm_gecko.c b/drivers/pwm/pwm_gecko.c
index 55c21a8681..cec2cb8a4b 100644
--- a/drivers/pwm/pwm_gecko.c
+++ b/drivers/pwm/pwm_gecko.c
@@ -11,6 +11,10 @@
 #include <em_cmu.h>
 #include <em_timer.h>
 
+#ifndef _TIMER_CC_CTRL_MODE_MASK
+#define _TIMER_CC_CTRL_MODE_MASK _TIMER_CFG_MODE_MASK
+#endif
+
 /** PWM configuration. */
 struct pwm_gecko_config {
 	TIMER_TypeDef *timer;
@@ -44,6 +48,11 @@ static int pwm_gecko_set_cycles(const struct device *dev, uint32_t channel,
 			(channel * _TIMER_ROUTELOC0_CC1LOC_SHIFT),
 			cfg->location << (channel * _TIMER_ROUTELOC0_CC1LOC_SHIFT));
 		BUS_RegMaskedSet(&cfg->timer->ROUTEPEN, 1 << channel);
+#elif defined(_GPIO_TIMER_ROUTEEN_MASK)
+  volatile uint32_t * route_register = &GPIO->TIMERROUTE[TIMER_NUM(cfg->timer)].CC0ROUTE;
+  route_register += cfg->channel;
+  *route_register = (cfg->port << _GPIO_TIMER_CC0ROUTE_PORT_SHIFT)
+                    | (cfg->pin << _GPIO_TIMER_CC0ROUTE_PIN_SHIFT);
 #else
 #error Unsupported device
 #endif
diff --git a/dts/arm/silabs/efr32mg24.dtsi b/dts/arm/silabs/efr32mg24.dtsi
index 5255f8e618..cb82bd25be 100644
--- a/dts/arm/silabs/efr32mg24.dtsi
+++ b/dts/arm/silabs/efr32mg24.dtsi
@@ -145,6 +145,18 @@
 			interrupts = <43 0>;
 			status = "disabled";
 		};
+
+		timer0: timer@50048000 {
+			compatible = "silabs,gecko-timers";
+			reg = <0x50048000 0x400>;
+			status = "disabled";
+
+			pwm {
+				compatible = "silabs,gecko-pwm";
+				status = "disabled";
+				#pwm-cells = <3>;
+			};
+		};
 	};
 };
 

```

fkokosinski commented:
Hi @tgorochowik, @mrfuchs.

I had to force push changes to address comments. PTAL at this PR. Thanks!

fkokosinski commented:
Hi @ihidchaos! I'm afraid I cannot help you, as we don't support PWM in this initial PR for this platform. Please feel free to create an issue for it after this PR is merged.

mrfuchs commented on dts/arm/silabs/efr32mg24b310f1536im48.dtsi:
```diff
@@ -0,0 +1,26 @@
+					 "silabs,efr32mg24", "silabs,efr32",
+					 "simple-bus";
+
+		flash-controller@50030000 {
+			flash0: flash@0 {
+				reg = <0 DT_SIZE_K(1536)>;
```
Without having the [FLASH_BASE_ADDRESS](https://docs.zephyrproject.org/latest/kconfig.html#CONFIG_FLASH_BASE_ADDRESS) set in [zephyr,flash](https://docs.zephyrproject.org/latest/build/dts/api/api.html), `west flash` (using the jlink runner) refuses to program the device for me due to the HEX file starting at address 0.

Shouldn't these lines consider the flash memory offset of 0x08000000 (cf. [figure 3.2. EFR32MG24 Memory Map](https://www.silabs.com/documents/public/data-sheets/efr32mg24-datasheet.pdf#page=20))? Changing the device tree like this allows me to successfully use `west flash`:

```suggestion
		flash-controller@50030000 {
			flash0: flash@8000000 {
				reg = <0x08000000 DT_SIZE_K(1536)>;
```

fkokosinski replied:
Hi, I can't seem to reproduce what you're describing here, i.e. west flashes the board normally without any modifications. Could you describe your env in more detail, e.g. OS, JLink ver, west ver etc., and paste the error message?

ZengYiiii commented:
Hi, my project uses EFR32BG24BxxxF1536, is there any support?
